### ISSUE TITLE
bring back async frame buffer reads

### ISSUE THREAD

**Issue by purplemarshmallow** [gonetz/GLideN64#679]
This feature was removed because it's not compatible with some games https://github.com/gonetz/GLideN64/commit/769ab7ac40bbbb0585df473363469a433fa47708
My  suggestion:
-disable "copy framebuffer to RDRAM" in the ini if a game does not need it
-bring async frame buffer reads back and enable it only for compatible games
This should improve performance in many games

**Comment by purplemarshmallow** [comment]
here's my async branch https://github.com/purplemarshmallow/GLideN64/tree/async
some testing:
compatible with async frame buffer reads:
Mario Kart, Perfect Dark, Banjo Kazooie, Glover2, Body Harvest, Castlevania, Donkey Kong 64, F-Zero X, Rayman 2, Shadow Man
not compatible with async frame buffer reads:
F-1 Pole Position, Pokemon Snap, Castlevania 2, Tonic Trouble, Bio F.R.E.A.K.S,

**Comment by AmbientMalice** [comment]
>-disable "copy framebuffer to RDRAM" in the ini if a game does not need it

For what it's worth, knowing this with 100% certainty is difficult. Also, I think it would be better to use async by default and disable it for games that are broken by it, because I suspect the games that are broken will be a minority.

**Comment by theboy181** [comment]
@purplemarshmallow 

Since I like to follow along but cant build this on my own you mind publishing this branch for me?


thanks in advance

**Comment by purplemarshmallow** [comment]
here's a build https://drive.google.com/file/d/0B7Y6r4SpC_QQdFZ6dDJNU01Wa1k/view?usp=sharing
every game that needs "copy framebuffer to RDRAM" option must be tested to see if it's compatible with async frame buffer reads

**Comment by AmbientMalice** [comment]
I have to install new Nvidia drivers because Win 10 killed the Nvidia control panel, meaning my GPU is stuck is power save mode, but Quake II seems significantly faster with async.

Unfortunately, there's still the broken textures and incorrect 32 bit colour buffer stuff. Out of curiosity, why doesn't GLideN64 do 32bit, and will implimenting it be a problem, performance-wise?

**Comment by purplemarshmallow** [comment]
Unfortunately it causes glitches in Quake II, the text and the HUD should not be visible in the pause menu.  I think it should be disabled for this game.

About the 32bit stuff, I think GLideN64 does not convert the framebuffer into the correct format before copying to RDRAM. Not sure how it can be fixed.
![gliden64_quake_ii_000](https://cloud.githubusercontent.com/assets/7283660/9427980/ff3668b6-4998-11e5-8ce6-ea00fe1289cb.jpg)
![gliden64_quake_ii_001](https://cloud.githubusercontent.com/assets/7283660/9427982/02f6dcec-4999-11e5-974b-f7df831161ac.jpg)

**Comment by gonetz** [comment]
The idea to return async frame buffer reads is generally right. I just did not find time for it.

What is incorrect with 32 bit colour buffer stuff in QII? Screen shots look correct.

**Comment by purplemarshmallow** [comment]
see #348 
The loading screen looks like this:
![gliden64_quake_ii_002](https://cloud.githubusercontent.com/assets/7283660/9428862/475b12bc-49bd-11e5-8e90-d79ac23cf544.jpg)

**Comment by AmbientMalice** [comment]
>Unfortunately it causes glitches in Quake II, the text and the HUD should not be visible in the pause menu. I think it should be disabled for this game.

That's unfortunate. I suppose that this could be one of those "causes minor issues in this particular game but enable it for a speed boost" cases, kinda like how Dolphin treats certain settings. But I suppose the complicating factor is that GLideN64 has its default settings in a file most people won't be able to access.

**Comment by AmbientMalice** [comment]
Any chance of bringing this back as a per-game setting, @purplemarshmallow?

**Comment by purplemarshmallow** [comment]
Qt reads and writes the ini I haven't got around yet to mess with Qt. Also I'm thinking of ways to optimize this that work in every game. One idea is to split up reads from the video card and writes to RDRAM and read asynchronously only once per frame. But I'm not sure if it can work out

**Comment by AmbientMalice** [comment]
@purplemarshmallow In the meantime, would you mind updating your async branch with the current method for me to do some speed tests? I tried updating it myself myself and got tripped up, likely because those fixes made to LLE a while back, where code was moved from between gDP.cpp and RSP.cpp Not really sure.

**Comment by purplemarshmallow** [comment]
I already uploaded a build here's the link again: https://drive.google.com/file/d/0B7Y6r4SpC_QQdFZ6dDJNU01Wa1k/view?usp=sharing Is this good enough? My branch now has a merge conflict

**Comment by AmbientMalice** [comment]
I'm aware of the merge conflict. I wasn't keen on comparing the old build against a new build since FB behavior has been juggled around since then, plus, IIRC, that build probably isn't compatible with the version number of the ini file anymore.

**Comment by gonetz** [comment]
I re-implemented async buffer read. Code in async_fb_read branch:
https://github.com/gonetz/GLideN64/commit/5fd2b6e5dee18a47f210da6f100fd32a03b90de1

Binary:
https://drive.google.com/file/d/0B0YqMPjGo3B2WDNZcWF5cnNQNTg/view?usp=sharing

Custom config settings file need to be corrected: games, which are compatible with async mode should have frameBufferEmulation\copyToRDRAM=2

**Comment by AmbientMalice** [comment]
On my PC, I'm looking at 550fps on Mario 64 using async and 450fps without. Depending on the game, the improvement can be minor or dramatic. @gonetz, I'm wondering how you would feel about making copyToRDRAM=2 the default, and then setting copyToRDRAM=1 for games that misbehave with it enabled. My assumption is that those games are very much in the minority.

There is one concern that occured to me -- when using async and also FB to texture, there would be a single frame of lag, wouldn't there? The frame rendered fullscreen would be one frame behind. That's something to think about since a lot of people hate lag. Maybe force sync mode if FB to texture is on or something.

**Comment by gonetz** [comment]
>how you would feel about making copyToRDRAM=2 the default, and then setting copyToRDRAM=1 for games that misbehave with it enabled.

Sounds good.

>when using async and also FB to texture, there would be a single frame of lag, wouldn't there?

In most cases CPU reads current content of frame buffer in RDRAM, post process it and FB to texture shows the result. That is fb read and fb to texture are in sync with each other. I tested ConkerBFD "It's war" cutscene and JFG rain  - both work the same in sync and async mode. There are games, where CPU just add new content to the frame, e.g. Dr.Mario, but such games do not need fb copy to RDRAM.

**Comment by purplemarshmallow** [comment]
>I tested ConkerBFD "It's war" cutscene and JFG rain

It's war cutscene is broken for me the red colors are incorrect. With async framebuffer reads it's broken even more. But I think these problems should be fixed with notifications from the emulator. Copy color probably needs to be reworked. JFG rain only works on mupen?

**Comment by AmbientMalice** [comment]
>JFG rain only works on mupen?

If it does, and I have no idea if it does, it's probably because copyFromRDRAM behaves wrong on the Zilmar version. (The jiggling image and incorrect colours and all that stuff.)

**Comment by gonetz** [comment]
>  It's war cutscene is broken for me the red colors are incorrect.

Yes, something wrong with it. But I don't see difference between sync and async modes here.

> JFG rain only works on mupen?

No. It works with Project64 as well.

> copyFromRDRAM behaves wrong on the Zilmar version
Does it work better with mupen64plus? Screenshots please.

**Comment by purplemarshmallow** [comment]
here are screenshots but the mupen64plus version does not have this problem: #603

**Comment by AmbientMalice** [comment]
@gonetz It's kinda hard to capture in screenshots. For me, Mario 64 with FB to texture and FB to RAM causes heavy up and down jiggling similar to interlacing errors. Quake II looks washed out and judders. I suspect this is somehow related to Quake II's 32 bit buffer problems.

**Comment by gonetz** [comment]
>For me, Mario 64 with FB to texture and FB to RAM causes heavy up and down jiggling similar to interlacing errors.

Yes, I see it. Need to be open as separate ticket.

**Comment by gonetz** [comment]
I want to merge "async_fb_read" branch with master. Do you know any regression caused by my changes?

**Comment by AmbientMalice** [comment]
@gonetz
>I want to merge "async_fb_read" branch with master. Do you know any regression caused by my changes?

Not that I'm aware of. Once you've merged, I'll make a PR to set games I know have issues to sync mode.

**Comment by purplemarshmallow** [comment]
I think sync is still default I'll agree with @AmbientMalice it's probably better to make async default and set the sync mode for the few games that need it. Again here's the testing I did some time ago:

>compatible with async frame buffer reads:
 Mario Kart, Perfect Dark, Banjo Kazooie, Glover2, Body Harvest, Castlevania, Donkey Kong 64, F-Zero X, Rayman 2, Shadow Man, Super Smash Bros, Mini Racers, Bakushou Jinsei 64, Clay Fighter - Sculptor's Cut, Conker's BFD, Donald Duck goin' quackers, Bomberman 64

> not compatible with async frame buffer reads:
 F-1 Pole Position, Pokemon Snap, Castlevania 2, Tonic Trouble, Bio F.R.E.A.K.S, Quake II, Heiwa Pachinko World, Hexen

**Comment by AmbientMalice** [comment]
DK64 has a similar problem to Quake II. HUD elements disappear one frame before menu appears. With async, HUD elements can be captured as part of blurred background.

There's a risk Perfect Dark has the same issue. Haven't tested yet.

**Comment by purplemarshmallow** [comment]
I'm not sure how to deal with these games. async causes minor inaccuracies in all games but mostly it's hardly noticeable. But hardly any game will look 100% perfect.

**Comment by AmbientMalice** [comment]
>I'm not sure how to deal with these games. async causes minor inaccuracies in all games but mostly it's hardly noticeable. But hardly any game will look 100% perfect.

TBH, the only people who are going to notice tiny, subtle bugs like the DK64 pause menu are obsessives like myself. I just think it's important to note these small inaccuracies when judging whether to force a game to use async or sync mode.

**Comment by gonetz** [comment]
>it's probably better to make async default and set the sync mode for the few games that need it.

Ok, let's make async default, but along with corrected custom ini.
